src/rules/no_shadow_restricted_names.rs

```diff
@@ -0,0 +1,286 @@
+// Copyright 2020 the Deno authors. All rights reserved. MIT license.
+use super::Context;
+use super::LintRule;
+use crate::scopes::{ScopeManager, ScopeVisitor};
+use swc_ecma_ast::{
+  ArrowExpr, AssignExpr, CatchClause, Expr, FnDecl, FnExpr, Ident, Module,
+  ObjectPatProp, Pat, PatOrExpr, VarDecl,
+};
+use swc_ecma_visit::{Node, Visit};
+
+pub struct NoShadowRestrictedNames;
+
+impl LintRule for NoShadowRestrictedNames {
+  fn new() -> Box<Self> {
+    Box::new(NoShadowRestrictedNames)
+  }
+
+  fn lint_module(&self, context: Context, module: Module) {
+    let mut scope_visitor = ScopeVisitor::new();
+    scope_visitor.visit_module(&module, &module);
+
+    let scope_manager = scope_visitor.consume();
+    let mut visitor =
+      NoShadowRestrictedNamesVisitor::new(context, scope_manager);
+    visitor.visit_module(&module, &module);
+  }
+
+  fn code(&self) -> &'static str {
+    "no-shadow-restricted-names"
+  }
+}
+
+#[allow(dead_code)]
+pub struct NoShadowRestrictedNamesVisitor {
+  context: Context,
+  scope_manager: ScopeManager,
+}
+
+impl NoShadowRestrictedNamesVisitor {
+  fn new(context: Context, scope_manager: ScopeManager) -> Self {
+    Self {
+      context,
+      scope_manager,
+    }
+  }
+
+  fn is_restricted_names(&self, ident: &Ident) -> bool {
+    match ident.sym.to_string().as_str() {
+      "undefined" | "NaN" | "Infinity" | "arguments" | "eval" => true,
+      _ => false,
+    }
+  }
+
+  fn check_pat(&self, pat: &Pat) {
+    match pat {
+      Pat::Ident(ident) => {
+        let scope = self.scope_manager.get_scope_for_span(ident.span);
+
+        if &ident.sym.to_string() == "undefined" {
+          if let Some(binding) = self
+            .scope_manager
+            .get_binding(&scope, ident.sym.to_string().as_str())
+          {
+            if binding.name == *ident.sym.to_string() {
+              self.report_shadowing(&ident);
+            }
+          }
+
+          return;
+        }
+
+        self.check_shadowing(ident);
+      }
+      Pat::Expr(expr) => {
+        if let Expr::Ident(ident) = expr.as_ref() {
+          self.check_shadowing(ident);
+        }
+      }
+      Pat::Array(array_pat) => {
+        for el in &array_pat.elems {
+          if el.is_some() {
+            self.check_pat(el.as_ref().unwrap())
+          }
+        }
+      }
+      Pat::Object(object_pat) => {
+        for prop in &object_pat.props {
+          match prop {
+            ObjectPatProp::Assign(assign) => {
+              self.check_shadowing(&assign.key);
+            }
+            ObjectPatProp::Rest(rest) => self.check_pat(&rest.arg),
+            ObjectPatProp::KeyValue(key_value) => {
+              self.check_pat(&key_value.value);
+            }
+          }
+        }
+      }
+      _ => {}
```

Comment:
it would be great if we could recognize the rest patterns like ```const [...undefined] = []```.
can you add `Rest` arm, please?

Reply:
Done, thanks @disizali. i also add more tests